tests/fe/fe_nedelec_singularity_01.cc

```diff
@@ -0,0 +1,798 @@
+
+                    for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                      {
+                        const Tensor<1, dim> phi_j =
+                          fe_values[electric_field].value(j, q);
+                        const Tensor<1, dim> curl_phi_j =
+                          fe_values[electric_field].curl(j, q);
+
+                        std::complex<double> matrix_sum = 0;
+
+                        matrix_sum +=
+                          std::pow(omega, 2) *
+                          (-epsilon_0 * mu_0 * epsilon_r * phi_i * phi_j);
+                        matrix_sum += omega * std::complex<double>(0, 1) *
+                                      mu_0 * sigma * phi_i * phi_j;
+                        matrix_sum += (1 / mu_r) * curl_phi_i * curl_phi_j;
+
+                        cell_matrix(i, j) += matrix_sum * fe_values.JxW(q);
+                      }
+
+                    cell_rhs(i) += phi_i * rhs * fe_values.JxW(q);
+                  }
+              }
+            cell->get_dof_indices(local_dof_indices);
+            constraints.distribute_local_to_global(cell_matrix,
+                                                   cell_rhs,
+                                                   local_dof_indices,
+                                                   system_matrix,
+                                                   system_rhs);
+          }
+      }
+    system_matrix.compress(VectorOperation::add);
+    system_rhs.compress(VectorOperation::add);
+  }
+
+
+
+  template <int dim, typename FiniteElementT>
+  void
+  NedelecSingularity<dim, FiniteElementT>::solve()
+  {
+    LinearAlgebraPETSc::MPI::Vector completely_distributed_solution(
+      locally_owned_dofs, mpi_communicator);
+
+    SolverControl                    solver_control;
+    PETScWrappers::SparseDirectMUMPS solver(solver_control, mpi_communicator);
+    solver.solve(system_matrix, completely_distributed_solution, system_rhs);
+
+    constraints.distribute(completely_distributed_solution);
+    locally_relevant_solution = completely_distributed_solution;
+  }
+
+
+
+  template <int dim, typename FiniteElementT>
+  void
+  NedelecSingularity<dim, FiniteElementT>::output_results()
+  {
+    {
+      const ExactSolution<dim> exact_solution_function;
+      Vector<double> difference_per_cell(triangulation.n_active_cells());
+
+      VectorTools::integrate_difference(dof_handler,
+                                        locally_relevant_solution,
+                                        exact_solution_function,
+                                        difference_per_cell,
+                                        QGauss<dim>(fe_order + 2),
+                                        VectorTools::L2_norm);
+      const double L2_error =
+        VectorTools::compute_global_error(triangulation,
+                                          difference_per_cell,
+                                          VectorTools::L2_norm);
```

Comment:
I copied this pattern from other tests. How does the `L2_norm` work? The result of `integrate_difference()` combined with `compute_global_error()` is `\int_Omega\sqrt(simulated_solution - exact_solution)^2` ? What does it happen if one cell is smaller than the rest of the cells. Does it take into account the size of the cell? The `L2_norm` is normalize by the total volume of the domain?

What does it happen if you run the test and the `L2_error` is smaller? The test would fail in the case of a smaller `L2_error`. It should always give roughly the same error because the mesh has the same size, right?

Reply:
Yes, it's the integral, so small cells are weighted less than large cells. (Each entry of `difference_per_cell` has a value of `sqrt(\int_K |u-u_h|^2)`. The `compute_global_error()` function then takes the sum of squares of these values, and takes the square root -- which corresponds to the global error: `sqrt(\int_\Omega |u-u_h|^2)`.)